### PR TITLE
[MV-1090] [MV-MIOpen] Port ResourceApplyAdam

### PR DESCRIPTION
**Pull request by hieule88**
* Base on Adam op
* Added Nesterov mode 
* Fixed bug in kernel code
* This newly added mode does not affect performance.

### CODE REVIEW

File: driver/adam_driver.hpp [hunk 1]
```diff
@@ -449,44 +486,45 @@ int AdamDriver<Tgpu, Tref, Tgrad>::RunForwardGPU()
...
+                                                weight_decay,
+                                                eps,
+                                                amsgrad,
+                                                maximize,
+                                                nesterov,
+                                                adamw,
+                                                gradScaleDesc,
+                                                grad_scale_ptr,
+                                                foundInfDesc,
+                                                found_inf_ptr);
```
Comment (long10024070) [inline driver/adam_driver.hpp:528]:
```suggestion
                                                found_inf_ptr);
           MIOPEN_THROW_IF(status != miopenStatusSuccess, ...
```

**Review by long10024070 — COMMENTED** [review]
Your PR is looking good. I only have a comments for you.

File: test/gtest/adam.hpp [hunk 2]
```diff
@@ -258,9 +266,9 @@ struct AdamTest : public ::testing::TestWithParam<AdamTestCase>
         double threshold = std::numeric_limits<Tp>::epsilon();
         auto error       = miopen::rms_range(ref_param, param);
 
-        EXPECT_TRUE(miopen::range_distance(ref_param) == miopen::range_distance(param));
-        EXPECT_TRUE(error < threshold * 10) << "Error output beyond tolerance Error:" << error
-                                            << ",  Thresholdx10: " << threshold * 10;
+        ASSERT_EQ(miopen::range_distance(ref_param), miopen::range_distance(param));
```
Comment (anhskrttt) [inline test/gtest/adam.hpp:269]: tsao phải đổi từ `EXPECT_TRUE()` thành `ASSERT_TRUE()` ở đây thế ạ? Mình phải dùng assert thay vì expect ở đoạn `RunTest()` ạ?

Comment (long10024070) [reply test/gtest/adam.hpp:269]: Because this is a check for critical errors, if this check is failed, all the below checks will be meaningless.

Comment (anhskrttt) [reply test/gtest/adam.hpp:269]: i got it. But is this usage a "must" or just a "nice-to-have"? `EXPECT_TRUE` still seems to be ok to use, right?

File: driver/adam_driver.hpp (hunk 1 above)
Comment (hieule88) [reply driver/adam_driver.hpp:528]: resolved

**Pull request closed without merging**